resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/v2/KubernetesSparkDependencyService.scala

```diff
@@ -0,0 +1,106 @@
+
+  /**
+   * Register an application with the dependency service, so that the driver pod can retrieve them
+   * when it runs.
+   *
+   * @param driverPodName Name of the driver pod.
+   * @param driverPodNamespace Namespace for the driver pod.
+   * @param jars Application jars to upload, compacted together in tar + gzip format. The tarball
+   *             should contain the jar files laid out in a flat hierarchy, without any directories.
+   *             We take a stream here to avoid holding these entirely in memory.
+   * @param files Application files to upload, compacted together in tar + gzip format. THe tarball
+   *              should contain the files laid out in a flat hierarchy, without any directories.
+   *              We take a stream here to avoid holding these entirely in memory.
+   * @param kubernetesCredentials These credentials are primarily used to monitor the progress of
+   *                              the application. When the application shuts down normally, shuts
+   *                              down abnormally and does not restart, or fails to start entirely,
+   *                              the data uploaded through this endpoint is cleared.
+   * @return A unique token that should be provided when retrieving these dependencies later.
+   */
+  @PUT
+  @Consumes(Array(MediaType.MULTIPART_FORM_DATA, MediaType.APPLICATION_JSON))
+  @Produces(Array(MediaType.TEXT_PLAIN))
+  @Path("/dependencies")
+  def uploadDependencies(
+      @QueryParam("driverPodName") driverPodName: String,
+      @QueryParam("driverPodNamespace") driverPodNamespace: String,
+      @FormDataParam("jars") jars: InputStream,
+      @FormDataParam("files") files: InputStream,
+      @FormDataParam("kubernetesCredentials") kubernetesCredentials: KubernetesCredentials)
+      : String
+
+  /**
+   * Download an application's jars. The jars are provided as a stream, where the stream's
+   * underlying data matches the stream that was uploaded in {@link uploadDependencies}.
+   */
+  @GET
+  @Consumes(Array(MediaType.APPLICATION_JSON))
+  @Produces(Array(MediaType.APPLICATION_OCTET_STREAM))
+  @Path("/dependencies/jars")
+  def downloadJars(
+      @HeaderParam("Authorization") applicationSecret: String): StreamingOutput
+
+  /**
+   * Download an application's files. The jars are provided as a stream, where the stream's
+   * underlying data matches the stream that was uploaded in {@link uploadDependencies}.
+   */
+  @GET
+  @Consumes(Array(MediaType.APPLICATION_JSON))
+  @Produces(Array(MediaType.APPLICATION_OCTET_STREAM))
+  @Path("/dependencies/files")
+  def downloadFiles(
+      @HeaderParam("Authorization") applicationSecret: String): StreamingOutput
+
+  /**
+   * Retrieve the Kubernetes credentials being used to monitor the Spark application.
+   */
+  @GET
+  @Consumes(Array(MediaType.APPLICATION_JSON))
+  @Produces(Array(MediaType.APPLICATION_JSON))
+  @Path("/dependencies/credentials")
+  def getKubernetesCredentials(
```

Comment:
The more I think about this the more I would prefer this to not be part of the API. The submitter can mount the credentials themselves as secrets or else expect the pod to have them in the local disk.